File: src/Chapter04GraphBFS/BipartitionDetection.java
package Chapter04GraphBFS;

import java.util.LinkedList;
import java.util.Queue;

public class BipartitionDetection {
    private Graph G;
    private boolean[] visited;
    private int[] colors;
    private boolean isBipartite = true;


    public BipartitionDetection(Graph g) {
        G = g;
        visited = new boolean[G.V()];
        colors = new int[G.V()];
        for (int i = 0; i < G.V(); i++) {
            colors[i] = -1;
        }
        for (int v = 0; v < G.V(); v++) {
            if (!visited[v]) {
                if (!bfs(v)) {
                    isBipartite = true;
                    break;
                }
            }
        }
    }

    public boolean bfs(int v) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(v);
        visited[v] = true;
        colors[v] = 0;
        while (!queue.isEmpty()) {
            int s = queue.remove();
            for (int i : G.adj(s)) {
                if (!visited[i]) {
                    queue.add(i);
                    visited[i] = true;
                    colors[i] = 1 - colors[s];
                } else {
                    if (colors[i] == colors[s]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public boolean isBipartite() {
        return isBipartite;
    }

    public static void main(String[] args) {

        Graph g = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g.txt");
        BipartitionDetection bipartitionDetection = new BipartitionDetection(g);
        System.out.println(bipartitionDetection.isBipartite);

        Graph g2 = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g2.txt");
        BipartitionDetection bipartitionDetection2 = new BipartitionDetection(g2);
        System.out.println(bipartitionDetection2.isBipartite);
    }
}

File: src/Chapter04GraphBFS/python/BipartitionDetection.py
from collections import deque
from Chapter02GraphDFS.python.adjSet import adjSet as Graph


class BiPartitionDetection:

    def __init__(self, G):
        self._G = G
        self._colors = [-1] * G.V
        self._is_bi_partition_graph = True

        for v in range(G.V):
            if self._colors[v] == -1:
                if not self._bfs(v):
                    self._is_bi_partition_graph = False
                    break

    def _bfs(self, s):
        queue = deque()
        queue.append(s)
        self._colors[s] = 0

        while queue:
            v = queue.popleft()
            for w in self._G.adj(v):
                if self._colors[w] == -1:
                    queue.append(w)
                    self._colors[w] = 1 - self._colors[v]
                # 如果下一个点w的颜色是当前处理点的颜色
                # 说明不能对其染色了
                # 即说明当前的图不是二分图
                elif self._colors[w] == self._colors[v]:
                    return False
        return True

    def is_bi_partition_graph(self):
        return self._is_bi_partition_graph

    def colors(self):
        return self._colors


if __name__ == '__main__':
    filename = '../g.txt'
    g = Graph(filename)
    bi_partition_detection = BiPartitionDetection(g)
    print('Is this a bi-partition graph? : {}'.format(
        bi_partition_detection.is_bi_partition_graph()),
    )

    filename = '../g2.txt'
    g = Graph(filename)
    bi_partition_detection = BiPartitionDetection(g)
    print('Is this a bi-partition graph? : {}'.format(
        bi_partition_detection.is_bi_partition_graph()),
    )

    filename = '../gg3.txt'
    g = Graph(filename)
    bi_partition_detection = BiPartitionDetection(g)
    print('Is this a bi-partition graph? : {}'.format(
        bi_partition_detection.is_bi_partition_graph()),
    )
